File: 프로그래머스/2/12946. 하노이의 탑/하노이의 탑.cpp
#include <string>
#include <vector>

using namespace std;

void hanoi(int n, int start, int mid, int end,
          vector<vector<int>>& answer) {
    
    // 1개일 때 start -> end
    if(n == 1){
        answer.push_back(vector<int>{start, end});
        return;
    }
    
    // n -1 개를 mid로
    hanoi(n - 1, start, end, mid, answer);
    
    // 1개 남은 것을 end로
    answer.push_back(vector<int>{start, end});
    
    // mid로 옮긴 n -1 개를 다시 end로
    hanoi(n - 1, mid, start, end, answer);
    
    
}

vector<vector<int>> solution(int n) {
    vector<vector<int>> answer;
    hanoi(n, 1, 2, 3, answer);
    return answer;
}

File: 프로그래머스/2/12946. 하노이의 탑/하노이의 탑.java
import java.util.*;
import java.util.stream.*;

class Solution {
    List<int[]> list = new ArrayList<>();
    public int[][] solution(int n) {
        hanoi(1,2,3,n);
        return list.stream()
            .toArray(int[][]::new);
    }
    
    void hanoi(int start, int mid, int from, int n){
        if(n == 1){
            System.out.println("" + start + " -> " + from);
            list.add(new int[]{start, from});
            return;
        }
        
        // n - 1개 중간에 옮기기
        hanoi(start, from, mid, n - 1);
        System.out.println("" + start + " -> " + from);
        list.add(new int[]{start, from});
        hanoi(mid, start, from, n - 1);
    }
}